lib/hammer_cli_foreman/api/connection.rb

```diff
@@ -30,18 +45,34 @@ def login_status
 
       protected
 
-      def create_authenticator(uri, settings)
+      # If the settings in foreman.yml has use_sessions as true
+      # and if there exists a session_file with valid contents, we use the auth_type from sessions_file
+      # Thus if the session expires (indicated by nil session_id), we use the
+      # same auth_type for re-authentication as was used by the previous session.
+      # Else we use the passed auth_type.
+      def conditioned_auth_type(auth_type, settings)
+        use_session = settings.get(:foreman, :use_sessions)
+        if use_session && !HammerCLIForeman.session_contents.empty? &&
+          HammerCLIForeman.session_contents[:session_id].nil?
+          HammerCLIForeman.session_contents[:auth_type]
+        else
+          auth_type
+        end
+      end
+
+      def create_authenticator(uri, settings, auth_type)
         return @authenticator if @authenticator
+        auth_type = conditioned_auth_type(auth_type, settings)
 
         if ssl_cert_authentication?(settings) && !use_basic_auth?(settings)
           @authenticator = VoidAuth.new(_('Using certificate authentication.'))
-        else
+        elsif auth_type == AUTH_TYPES[:basic_auth]
           if settings.get(:foreman, :use_sessions)
             @authenticator = InteractiveBasicAuth.new(
               settings.get(:_params, :username) || ENV['FOREMAN_USERNAME'],
               settings.get(:_params, :password) || ENV['FOREMAN_PASSWORD']
             )
-            @authenticator = SessionAuthenticatorWrapper.new(@authenticator, uri)
+            @authenticator = SessionAuthenticatorWrapper.new(@authenticator, uri, auth_type)
           else
             username = settings.get(:_params, :username) || ENV['FOREMAN_USERNAME'] || settings.get(:foreman, :username)
             password = settings.get(:_params, :password) || ENV['FOREMAN_PASSWORD']
```

Comment:
From what I have understood is that a user might be logged in through `settings.get(:_params, :username) || ENV['FOREMAN_USERNAME']`. Now in this case, if the password is not provided then it should be asked interactively to the user and not taken from `settings.get(:foreman, :password)`.

There will always be a user and password by default in the settings file, so we want to make sure that if we are using a user from the settings file only then we use the password from the settings file. 

If we do something like adding the condition `(settings.get(:foreman, :password)` to the above line then it might happen that a user from `settings.get(:_params, :username) || ENV['FOREMAN_USERNAME']` catch the password from the settings which would be wrong.

Let me know your thoughts on this.